Review Button alignment, grouping, and order docs

jorytindall opened the pull request:
### :pushpin: Summary

If merged this PR will make visible the documentation and guidelines around button alignment, grouping and order. It's ready for a first round review.

This is likely low-priority documentation (for now), but would be good to get an initial review around it to get a v1 available for consumers.

[Preview here](https://hds-website-25esn7w0m-hashicorp.vercel.app/patterns/button-alignment)

I haven't added the sidebar back into the patterns page, but have done so in [HDS-1220](https://github.com/hashicorp/design-system/pull/1220).

### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: [HDS-1593](https://hashicorp.atlassian.net/browse/HDS-1593)


[HDS-1593]: https://hashicorp.atlassian.net/browse/HDS-1593?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[HDS-1220]: https://hashicorp.atlassian.net/browse/HDS-1220?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

didoo commented on website/docs/patterns/button-alignment/index.md:
```diff
@@ -2,7 +2,7 @@
 title: Button alignment
 description: Guidelines for aligning, ordering, and grouping action buttons.
 navigation:
-  hidden: true
+  hidden: false
```
suggestion: instead of `false` let's just get rid of the `hidden` argument (and at that point also the `navigation` one)

jorytindall replied:
Hmmmm, that's a good point, I didn't think about that since the content was merged already. 

I'll remove the hidden for now. For content review purposes I'm not sure what makes sense, but should I remove the content in a new PR and add it back in a subsequent PR?

didoo replied:
No idea :) 

@alex-ju git guru, any ideas/suggestion how we can resolve this?

didoo replied:
@jorytindall nne option would be (@alex-ju what do you think?) to do two sequential commits in this PR, one in which you delete the markdown files for the documentation and one in which you reverse the commit (essentially adding back the files). In this way we could comment on the "revert" commit directly (it's possible, afaik).
At the end of the process, we squash all the commits in a single one, and it's like nothing has happened.

alex-ju replied:
Let me see if I got this right. The documentation was added in #1092, but was hidden at the time. In this PR we make it visible and remove some exceptions. However we'd like to re-review the content, so a visual diff in GitHub would be necessary.

First of all, merging the branch 'main' into jt/review-button-alignment didn't help as it blocked us from being able to rebase. So I got rid of that merge first so that we can operate on this branch. This means you'd have to reset the head before operating further on this branch, Jory. So please run the following command on this branch, on your local machine, @jorytindall: `git reset --hard origin/jt/review-button-alignment` – this will make your local setup match what you see now in GitHub and will allow you to add further commits as needed based on feedback.

Didoo's suggestion sounds good to me. So this is what I did.
The original PR had 16 commits, so quite difficult to revert those, but we can use the merge commit instead (which sums up the changes). Because it is a merge commit it requires some special attention, so I did `git revert 675e257 -m 1` (if it wasn't a regular commit `git revert 675e257` would've been enough). Then I reverted the revert commit so that we get the diff we wanted and reviewers can now make suggestions on the content.

Hope that helps.

jorytindall replied:
Whoa, super helpful, thanks @alex-ju !

didoo reviewed (commented):
@jorytindall in this PR I see only the deletion of two files and the change of the `hidden` value. are we also supposed to review the actual content of the page? In that case, not sure how/where we can add comment, since it's not in this PR... 😕

didoo commented on website/docs/patterns/button-alignment/partials/overview/overview.md:
```diff
@@ -0,0 +1,92 @@
+
+## Constrained vs. unconstrained
+
+**Constrained** refers to a component or pattern that is wrapped within a container element, while **unconstrained** refers to components that exist at the page level.
```
Maybe "The term **constrained** refers to..."?

didoo commented on website/docs/patterns/button-alignment/partials/overview/overview.md:
```diff
@@ -0,0 +1,92 @@
+
+The Gutenberg diagram refers to a pattern the eyes follow when moving or scanning through **evenly distributed, homogenous content.** This pattern is most relevant in text-heavy scenarios like novels and newspapers, but is relevant in a UX context depending on the complexity of the content.
+
+![Visual example of a Gutenberg diagram](/assets/patterns/button-alignment/gutenberg-diagram.png)
```
This single image weights almost 600Kb 😬 (it's not optimized!). 
Some ideas:
- let's use JPG as format, with medium quality, if there are patterned gradients (like in this one)
- let's not use patterned gradients at all, for this kind of images (this would allow us to use flat PNGs, if not even SVGs in some cases) - personally I don't think they add much 
- whatever format we use, let's compress the images (be them vectors like SVGs or bitmap like PNGs or JPGs)

jorytindall replied:
Opted to remove the background and optimized the images. These were all from the presentation which I more or less ported over, but doesn't make sense for the current visual direction of the website

didoo commented on website/docs/patterns/button-alignment/partials/specifications/order.md:
```diff
@@ -0,0 +1,77 @@
+
+!!! Do
+
+![Destructive action order](/assets/patterns/button-alignment/order-destructive-action.png)
```
in the example just above, the image is out of the `Do/Dont` box, here inside. maybe use a similar approach for both?

didoo reviewed (changes requested):
I have added a comment about images, that I think is a blocker, plus other comments here and there.

Beyond that, a couple of general consideration: 
- do we really need to split the content in tabs? if yes, keep in mind we're not bound to use the same tab names of the component pages: if we want, we can have completely different tab labels, and organization of the content in tabs.
- the whole content of the first tab is around reading patterns, the word "button" is mentioned only at the very beginning, in the intro, and then no more (will appear only if a user clicks "Specifications"); could it be confusing for the reader, expecting to read about "buttons alignment, grouping and order"?
  - this for me is also a "smell" that maybe having all the content in a single page would work better for the patterns pages?
  - an alternative would be to change the "information architecture": first show the buttons patterns, and refer to the reading patterns as a drill-down information section; but also this solution is not perfect :(

jorytindall commented:
> * but also this solution is not perfect :(

Yeah, I agree, the information architecture could use some work but there isn't a 100% "right" solution. The reason why I led with the higher level strategy information (reading patterns, etc) is because understanding those concepts is key to the decision-making process for the entire pattern. Let me explore a little bit more and bump some things around.

heatherlarsen reviewed (commented):
I took a quick scan through the content, and have some high-level feedback:

- "Button alignment" as the title feels like it's only capturing one piece of the puzzle. Why not align this more closely to the `ButtonSet` component by calling this page "Button set"?
- Having all of the details about reading patterns on the first tab feels a little overwhelming and doesn't provide tactical information for consumers. This information would be more appropriate for users interested in diving deeper into the topic, but shouldn't be required reading to know which button alignment and ordering to use.

heatherlarsen commented on website/docs/patterns/button-alignment/partials/specifications/alignment.md:
```diff
@@ -0,0 +1,25 @@
+
+- **Left (start) alignment:** align actions to the left of the container or page.
+- **Right (end) alignment:**  align actions to the right of the container or page.
+- **Center alignment:** center actions in the middle of the container or page.
```
Do you have visual examples of when someone _might_ use either of these situations?

heatherlarsen commented on website/docs/patterns/button-alignment/partials/specifications/grouping.md:
```diff
@@ -0,0 +1,13 @@
+## Grouping
+
+How actions are grouped can influence what the user perceives as the primary action. Actions grouped correctly can alleviate decision fatigue and help users better understand the outcome of the action they take.
```
Suggestion: 
How actions are grouped can influence what the user perceives as the primary action. Actions grouped correctly can alleviate decision fatigue and help users better understand the outcome of their actions.

heatherlarsen reviewed (commented):
I left some grammar suggestions; sorry, I couldn't do them as normal suggestions.

jorytindall commented:
@heatherlarsen (and @didoo for visibility) updated following:

- Heather's grammar and prose feedback
- Reorganized the alignment section and added images
- Put "Guidelines" first and followed with "Research" which contains all of the reading patterns and related (but deep dive) content
- Renamed the docs to "Button organization" which seems to encompass alignment, order, and grouping fairly well (better than just calling it "Button alignment")

heatherlarsen commented on website/docs/patterns/button-organization/partials/guidelines/order.md:
```diff
@@ -0,0 +1,78 @@
+
+![Suboptimal grouping](/assets/patterns/button-organization/suboptimal-grouping.png)
+
+- User edits the settings of a cluster
+    - "Would you like to learn more about this action?" "No"
+    - "Would you like to cancel?" "No"
+    - "Would you like to save the changes to this cluster?" "Yes, finally, that’s what I was trying to do all along."
```
It doesn't feel like this should be a bullet point with nested bullet points since there's only one item at the top level. Have you considered something like:

> User edits the settings of a cluster:
> 
> - "Would you like to learn more about this action?" "No"
> - "Would you like to cancel?" "No"
> - "Would you like to save the changes to this cluster?" "Yes, finally, that's what I was trying to do all along."

heatherlarsen commented on website/docs/patterns/button-organization/partials/guidelines/order.md:
```diff
@@ -0,0 +1,78 @@
+#### Natural conversation based on priority
+![Optimal order](/assets/patterns/button-organization/optimal-ordering.png)
+
+- User edits the settings of a cluster
+    - "Would you like to save the changes to this cluster?" "Yes, thank you."
```
Same note about the bullet point with nested bullet point here too

heatherlarsen reviewed (changes requested):
I made one blocking comment about one of the guidelines we can discuss further, if necessary and a few other minor suggestions, but this looks great! ✨

jorytindall commented on website/docs/patterns/button-organization/partials/guidelines/alignment.md:
```diff
@@ -0,0 +1,47 @@
+
+!!! Dont
+
+![Center alignment example](/assets/patterns/button-organization/using-center-alignment.png)
```
Agreed, in the dropdown that alignment makes sense, I'll update the example to something different

heatherlarsen commented on website/docs/patterns/button-organization/partials/guidelines/order.md:
```diff
@@ -0,0 +1,80 @@
+
+User edits the settings of a cluster:
+
+"Would you like to save the changes to this cluster?" "Yes, thank you."
```
```suggestion
- "Would you like to save the changes to this cluster?" "Yes, thank you."
```

heatherlarsen commented on website/docs/patterns/button-organization/partials/guidelines/order.md:
```diff
@@ -0,0 +1,80 @@
+
+User edits the settings of a cluster:
+
+"Would you like to save the changes to this cluster?" "Yes, thank you."
```
```suggestion
- "Would you like to save the changes to this cluster?" "Yes, thank you."
```

jorytindall replied:
I specifically didn't bullet that because it's a single item. I don't have a strong opinion, happy to update it, I imagine the consistency you're talking about is consistency with the previous "dont" block that bullets multiple items

heatherlarsen replied:
I don't know if it's correct grammatically, but I always bullet these because it appears "nested" under what it's supposed to be an example of instead of a new thought. Maybe @KristinLBradley knows?

heatherlarsen reviewed (approved):
Looks great! Found one minor nit-picky consistency thing, but otherwise I'd consider this good to go ✨

MelSumner commented on website/docs/patterns/button-alignment/partials/accessibility/accessibility.md:
did this file get replaced @jorytindall ?
